pkg/resourcemanager/controller/health/add.go

```diff
@@ -67,19 +69,63 @@ type ControllerConfig struct {
 // AddToManagerWithOptions adds the controller to a Manager with the given config.
 func AddToManagerWithOptions(mgr manager.Manager, conf ControllerConfig) error {
 	// setup main health reconciler
+	healthReconciler := &reconciler{
+		syncPeriod:   conf.SyncPeriod,
+		classFilter:  &conf.ClassFilter,
+		targetClient: conf.TargetCluster.GetClient(),
+		targetScheme: conf.TargetCluster.GetScheme(),
+	}
+
 	healthController, err := controller.New(ControllerName, mgr, controller.Options{
 		MaxConcurrentReconciles: conf.MaxConcurrentWorkers,
-		Reconciler: &reconciler{
-			syncPeriod:   conf.SyncPeriod,
-			classFilter:  &conf.ClassFilter,
-			targetClient: conf.TargetCluster.GetClient(),
-			targetScheme: conf.TargetCluster.GetScheme(),
-		},
-		RecoverPanic: true,
+		Reconciler:              healthReconciler,
+		RecoverPanic:            true,
 	})
 	if err != nil {
 		return fmt.Errorf("unable to setup health reconciler: %w", err)
 	}
+	healthLogger := healthController.GetLogger()
+
+	if conf.TargetCacheDisabled {
+		// if the target cache is disable, we don't want to start additional informers
+		healthReconciler.EnsureWatchForGVK = func(gvk schema.GroupVersionKind, obj client.Object) error {
+			return nil
+		}
+	} else {
+		watchedObjectGVKs := sync.Map{}
+		healthReconciler.EnsureWatchForGVK = func(gvk schema.GroupVersionKind, obj client.Object) error {
+			// check if we have already added watch for GVK
+			// if not, store GVK in map
+			if _, ok := watchedObjectGVKs.LoadOrStore(gvk, nil); ok {
```

Comment:
Shouldn't we only store in case the below `Watch()` call succeeded?

Reply:
Definitely. I must have forgotten this when experimenting with `sync.Map` 🙈

Reply:
I returned to a pure `map+RWMutex` implementation (now I found the issue I had when originally working with it).
This should now be performant enough but also properly handle errors.